src/occa/internal/io/utils.cpp

```diff
@@ -442,5 +442,65 @@ namespace occa {
       fsync(fileno(fp));
       fclose(fp);
     }
+
+    std::string tmpFilenameBelow(const std::string &filename) {
+#if (OCCA_OS & (OCCA_LINUX_OS | OCCA_MACOS_OS))
+
+      /*
+        Return a temporary filename with the same basename as filename,
+        but in a temporary subdirectory.
+
+        This seems convoluted (and it is), but many variants to create
+        temporary file(name)s are unsafe, and we cannot use mkstemp
+        because it returns a file handle, not a file name, and we need
+        to able to pass the filename to e.g. compiler invocations as a string.
+
+        Unfortunate side effect: we have an extra directory that needs
+        to be cleaned. This is handled in the renameTmpFile function, so use
+        these in pairs.
+      */
+      const std::string expFilename = io::expandFilename(filename);
+      sys::mkpath(dirname(expFilename));
+      const std::string template_ = dirname(expFilename) + "XXXXXX";
+      size_t bufferSize = template_.size();
+      char buffer[bufferSize + 1];
+      ::memcpy(buffer, template_.c_str(), bufferSize);
+      // Set null terminator
+      buffer[bufferSize] = '\0';
+      char *dirname = mkdtemp(buffer);
+      OCCA_ERROR("Failed to create temporary directory for template [" << template_ << "]: " << strerror(errno),
+                 dirname != NULL);
+      return std::string(dirname) + "/" + std::string(basename(expFilename));
+#else
+#error "TODO: What to use on Windows?"
+#endif
+    }
+
+    void renameTmpFile(const std::string &filenameIn,
+                       const std::string &filenameOut) {
+      const std::string expFilename = io::expandFilename(filenameOut);
+      sys::mkpath(dirname(expFilename));
+
+      int status;
+      status = std::rename(filenameIn.c_str(), expFilename.c_str());
+      /*
+        On NFS filesystems, you can not assume that if the operation
+        failed, the file was not renamed. If the server does the rename
+        operation and then crashes, the retransmitted RPC which will be
+        processed when the server is up again causes a failure.
+      */
+      OCCA_ERROR("Failed to rename [" << filenameIn << "] to [" << expFilename << "]: " << strerror(errno),
+                 status == 0 || io::isFile(filenameOut));
+
+      /*
+        Assume the temporary file has been created using the tmpFilenameBelow
+        function, so we need to clean up the temporary directory as well.
+      */
+      const std::string tmpDirname = dirname(io::expandFilename(filenameIn));
+      status = rmdir(tmpDirname.c_str());
+      OCCA_ERROR("Failed to remove temporary directory [" << tmpDirname << "]: " << strerror(errno),
+                 status == 0);
```

Comment:
If there's no way in which the directory gets populated with other files, maybe the directory doesn't need to be created (it's just more precious metadata on a parallel filesystem).

Reply:
I would also prefer not to have a directory but a safe filename instead (to go into the base cache dir), but I didn't see a way to get one reliably. See my comment at the top of `tmpFilenameBelow`. If you have a better way I'll gladly rewrite this bit.

Reply:
So if there was an active adversary, they could observe the directory being created and then plop stuff in there. I don't think the compiler output file can be made robust to that (because you can't make the compiler use `mkstemp` internally and report back the name), so I'd instead focus on creating file names that don't collide. A hash of the content (or a cheap non-repeating number) plus a per-process salt (which could be strong, as in https://en.cppreference.com/w/cpp/numeric/random/random_device) would be more secure and debuggable.

Reply:
From the security perspective, this temp dir will always be below the occa cache dir, so any adversary could then probably also just populate the cache with malicious binaries directly?

Since this temporary file name is often used to capture the output I don't see how we can consistently use a hash based on that.

Naively I would try to maybe use something like generating a hash from (hashdir + pid + time + random number) to use as a filename?

Reply:
I mean that you could hash the input, such that the filename would reflect changes (or not), which might be useful when debugging. But anything with enough uniqueness would be fine.

Reply:
We are already inside a hashDir, which is based on the contents of the kernel file. So maybe we could also look at the contents of the intermediate files that we are using, but it sounds expensive?

The PID seems to be hard to come by in a platform-independent way, so I'm just experimenting with a hash based on time + random at the moment. Code compiles and runs, but I'm not actually stress-testing it at the moment.